File: packages/autonavi_location_flutter/ios/Classes/LocationStreamHandler.swift
import Flutter

class LocationStreamHandler: NSObject, FlutterStreamHandler {

    private let adapter = AMapLocationAdapter()
    private var eventSink: FlutterEventSink?

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        adapter.onLocation = { [weak self] result in self?.eventSink?(result) }
        adapter.onError = { [weak self] err in
            self?.eventSink?(FlutterError(code: err.code, message: err.message, details: nil))
        }
        adapter.startContinuous(options: parseOptions(arguments as? [String: Any] ?? [:]))
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        adapter.stop()
        eventSink = nil
        return nil
    }

    // MARK: - Public

    func getOnce(options: [String: Any], callback: @escaping (Any?) -> Void) {
        adapter.startOnce(options: parseOptions(options)) { result, err in
            if let err = err {
                callback(FlutterError(code: err.code, message: err.message, details: nil))
            } else {
                callback(result)
            }
        }
    }

    func updateOptions(_ options: [String: Any]) {
        adapter.updateOptions(parseOptions(options))
    }

    // MARK: - Private

    private func parseOptions(_ options: [String: Any]) -> LocationOptions {
        return LocationOptions(
            accuracy: options["accuracy"] as? Int ?? 2,
            intervalMs: options["intervalMs"] as? Int ?? 2000,
            needAddress: options["needAddress"] as? Bool ?? true
        )
    }
}

File: packages/autonavi_location_flutter/android/src/main/kotlin/com/autonavi/location/flutter/LocationStreamHandler.kt
package com.autonavi.location.flutter

import android.content.Context
import io.flutter.plugin.common.EventChannel

class LocationStreamHandler(private val context: Context) : EventChannel.StreamHandler {

    private val adapter = AMapLocationAdapter(context)

    override fun onListen(args: Any?, events: EventChannel.EventSink) {
        @Suppress("UNCHECKED_CAST")
        val options = parseOptions(args as? Map<String, Any> ?: emptyMap())
        adapter.onLocation = { result -> events.success(result) }
        adapter.onError    = { code, msg -> events.error(code, msg, null) }
        adapter.startContinuous(options)
    }

    override fun onCancel(args: Any?) {
        adapter.stop()
    }

    fun getOnce(
        options: Map<String, Any>,
        onSuccess: (Map<String, Any?>) -> Unit,
        onError: (String, String?) -> Unit,
    ) {
        adapter.startOnce(parseOptions(options), onSuccess, onError)
    }

    fun updateOptions(options: Map<String, Any>) {
        adapter.updateOptions(parseOptions(options))
    }

    fun dispose() {
        adapter.stop()
    }

    private fun parseOptions(options: Map<String, Any>): LocationOptions = LocationOptions(
        accuracy       = (options["accuracy"]       as? Int)    ?: 2,
        intervalMs     = (options["intervalMs"]     as? Int)    ?: 2000,
        needAddress    = (options["needAddress"]    as? Boolean) ?: true,
        distanceFilter = (options["distanceFilter"] as? Number)?.toFloat() ?: 0f,
    )
}
